### PR TITLE
Bug 1783088, update Hard pruning the registry procedure

### PR DESCRIPTION
**Pull request by ahardin-rh**
https://bugzilla.redhat.com/show_bug.cgi?id=1783088

Preview Build: http://file.rdu.redhat.com/~ahardin/Feb212020/hard-pruning-registry/applications/pruning-objects.html#pruning-hard-pruning-registry_pruning-objects

### CODE REVIEW

**Comment by ahardin-rh** [comment]
@dmage Thanks for your guidance. Please review and let me know if I understood your suggestions correctly. Thanks again!

**Comment by dmage** [comment]
/retitle Bug 1783088: update Hard pruning the registry procedure

I'll try to setup a cluster tomorrow and try to run the hard pruner on 4.x (master).

**Comment by dmage** [comment]
eh... I see

/retitle Bug 1783088, update Hard pruning the registry procedure

File: modules/pruning-hard-pruning-registry.adoc [hunk 1]
```diff
@@ -95,7 +89,7 @@ permissions in order to list some resources.
 ----
 $ service_account=$(oc get -n default \
     -o jsonpath=$'system:serviceaccount:{.metadata.namespace}:{.spec.template.spec.serviceAccountName}\n' \
-    dc/docker-registry)
+    dc/openshift-image-registry
 ----
 
 .. Add the `system:image-pruner` cluster role to the service account:
```
Comment (dmage) [inline modules/pruning-hard-pruning-registry.adoc]:
```
oc adm policy add-cluster-role-to-user system:image-pruner -z $service_account -n openshift-image-registry
```

File: modules/pruning-hard-pruning-registry.adoc [hunk 2]
```diff
@@ -113,26 +107,26 @@ changes are actually made:
 +
 ----
 $ oc -n default \
-    exec -i -t "$(oc -n default get pods -l deploymentconfig=docker-registry \
```
Comment (dmage) [inline modules/pruning-hard-pruning-registry.adoc]:
```
$ oc -n openshift-image-registry rsh deploy/image-registry /usr/bin/dockerregistry -prune=check
```

File: modules/pruning-hard-pruning-registry.adoc [hunk 3]
```diff
@@ -113,26 +107,26 @@ changes are actually made:
...
-    -- /usr/bin/dockerregistry -prune=check
+    -- /usr/bin/openshift-image-registry -prune=check
 ----
 +
 Alternatively, to get the exact paths for the prune candidates, increase the
 logging level:
 +
 ----
 $ oc -n default \
-    exec "$(oc -n default get pods -l deploymentconfig=docker-registry \
```
Comment (dmage) [inline modules/pruning-hard-pruning-registry.adoc]:
```
$ oc -n openshift-image-registry rsh deploy/image-registry env REGISTRY_LOG_LEVEL=info /usr/bin/dockerregistry -prune=check
```

File: modules/pruning-hard-pruning-registry.adoc [hunk 4]
```diff
@@ -154,14 +148,14 @@ pod to run the hard prune:
 +
 ----
 $ oc -n default \
-    exec -i -t "$(oc -n default get pods -l deploymentconfig=docker-registry -o jsonpath=$'{.items[0].metadata.name}\n')" \
```
Comment (dmage) [inline modules/pruning-hard-pruning-registry.adoc]:
```
$ oc -n openshift-image-registry rsh deploy/image-registry /usr/bin/dockerregistry -prune=delete
```

File: modules/pruning-hard-pruning-registry.adoc [hunk 5]
```diff
@@ -154,14 +148,14 @@ pod to run the hard prune:
...
-    -- /usr/bin/dockerregistry -prune=delete
+    exec -i -t "$(oc -n default get pods -l deploymentconfig=openshift-image-registry -o jsonpath=$'{.items[0].metadata.name}\n')" \
+    -- /usr/bin/openshift-image-registry -prune=delete
 ----
 +
 .Sample output
 ----
 $ oc exec docker-registry-3-vhndw \
-    -- /usr/bin/dockerregistry -prune=delete
+    -- /usr/bin/openshift-image-registry -prune=delete
```
Comment (dmage) [inline modules/pruning-hard-pruning-registry.adoc]: it's still `/usr/bin/dockerregistry`

File: modules/pruning-hard-pruning-registry.adoc [hunk 6]
```diff
@@ -173,5 +167,5 @@ After the prune is finished, the registry can be switched back to read-write
 mode by executing:
 +
 ----
-$ oc set env -n default dc/docker-registry REGISTRY_STORAGE_MAINTENANCE_READONLY-
+$ oc set env -n default dc//usr/bin/openshift-image-registry REGISTRY_STORAGE_MAINTENANCE_READONLY-
```
Comment (dmage) [inline modules/pruning-hard-pruning-registry.adoc]:
```
In `configs.imageregistry.operator.openshift.io/cluster`, set `spec.readOnly` to false.
```

File: modules/pruning-images.adoc [hunk 7]
```diff
@@ -117,7 +117,7 @@ minutes ago and is not currently referenced by:
 - any running Pods.
 - any pending Pods.
 - any ReplicationControllers.
-- any DeploymentConfigs.
+- any Deployments.
```
Comment (dmage) [inline modules/pruning-images.adoc]: you can keep DeploymentConfigs in this list, and add ReplicaSets.

File: modules/pruning-images.adoc [hunk 8]
```diff
@@ -128,7 +128,7 @@ the same project and is not currently referenced by:
...
```
Comment (dmage) [inline modules/pruning-images.adoc]: you can keep DeploymentConfigs in this list, and add ReplicaSets.

**Comment by ahardin-rh** [comment]
@dmage Thank you for your guidance! I made the updates. Please review and let me know if I missed anything. Preview build: http://file.rdu.redhat.com/~ahardin/Feb212020/hard-pruning-registry/applications/pruning-objects.html#pruning-hard-pruning-registry_pruning-objects

File: modules/pruning-hard-pruning-registry.adoc [hunk 9]
```diff
@@ -93,17 +77,17 @@ permissions in order to list some resources.
...
 
 .. Add the `system:image-pruner` cluster role to the service account:
 +
 ----
 $ oc adm policy add-cluster-role-to-user \
-    system:image-pruner \
-    ${service_account}
+    system:image-pruner -z \
+    ${service_account} \
+    $service_account -n openshift-image-registry
```
Comment (xiuwang) [inline modules/pruning-hard-pruning-registry.adoc]: Typo `$service_account` will be treat as user name whose not exist in the cmd

File: modules/pruning-hard-pruning-registry.adoc [hunk 10]
```diff
@@ -112,26 +96,23 @@ To see how many blobs would be removed, run the hard pruner in dry-run mode. No
...
-    -c 'REGISTRY_LOG_LEVEL=info /usr/bin/dockerregistry -prune=check'
+$ oc -n openshift-image-registry \
+    rsh deploy/image-registry env REGISTRY_LOG_LEVEL=info \
+    /usr/bin/dockerregistry -prune=check
 ----
 +
 .Truncated sample output
 ----
-$ oc exec docker-registry-3-vhndw \
+$ oc exec openshift-image-registry-3-vhndw \
```
Comment (xiuwang) [inline modules/pruning-hard-pruning-registry.adoc]: the image-registry pod is start with prefix `image-registry`, no openshift in the pod name

**Comment by xiuwang** [comment]
@ahardin-rh I Found other 'docker' words need updated in the original docs, such as:
1. Pushing to the registry initiated by `docker` daemon failures 
2. Execute the following command inside one running instance of a `docker-registry` pod to run the hard prune
3. oc exec `docker-registry-3-vhndw`   -- /usr/bin/dockerregistry -prune=delete

**Comment by xiuwang** [comment]
@dmage @ahardin-rh 
In the `Procedure` first step `log in with a user`, but in fourth step `Add the system:image-pruner role`, add image-pruner role to sa of openshift-image-registry. It is a little confused.
It's better to create a sa in the user's project, and added system:image-pruner role. But there is another potential mistake in doing this.
``` oc exec -n openshift-image-registry image-registry-666cf7dd8d-nvmnr     -- /bin/sh -c 'REGISTRY_LOG_LEVEL=info /usr/bin/dockerregistry -prune=check'  
Error from server (Forbidden): pods "image-registry-666cf7dd8d-nvmnr" is forbidden: User "xiuwang2" cannot create resource "pods/exec" in API group "" in the namespace "openshift-image-registry"```

**Comment by dmage** [comment]
it should be kubeadmin or another privileged user that has access to the openshift-image-registry namespace

**Comment by ahardin-rh** [comment]
@dmage @xiuwang Thank you for your comments. This is now updated. Please let me know if I missed anything. Thanks again!

**Comment by xiuwang** [comment]
LGTM

**Comment by xiuwang** [comment]
@ahardin-rh Hey, Ashley
I am sorry to miss to go through the entire doc `pruning-images.adoc` .
Now several things need updated too in this doc.
1.  Pruning images with the --namespace flag does not remove images, only `image streams`
    `image streams` need updated to `imagestreams`
2. Keeping up to three tag revisions, and keeping resources (images, `image streams` and Pods) younger than sixty minutes
    same as above
3. Four `oadm prune images` commands need update to `oc adm prune images`

**Comment by xiuwang** [comment]
@dmage  For the doc `pruning-images.adoc`, I have one question:
How to understand `the registry must be configured with storage:delete:enabled set to true`? If that means set `readOnly` in the configure instance.
Could you give me more guide about this?
Thanks

**Comment by dmage** [comment]
@xiuwang `storage:delete:enabled` is aways set to true in 4.x, so this statement can be removed.

**Comment by xiuwang** [comment]
> @ahardin-rh Hey, Ashley
> I am sorry to miss to go through the entire doc `pruning-images.adoc` .
> Now several things need updated too in this doc.
> 
> 1. Pruning images with the --namespace flag does not remove images, only `image streams`
>    `image streams` need updated to `imagestreams`
> 2. Keeping up to three tag revisions, and keeping resources (images, `image streams` and Pods) younger than sixty minutes
>    same as above
> 3. Four `oadm prune images` commands need update to `oc adm prune images`
@ahardin-rh Please also remove the sentence `For this operation to work properly, the registry must be configured with storage:delete:enabled set to true.`

**Comment by ahardin-rh** [comment]
@xiuwang Thank you! The updates are complete.

**Comment by xiuwang** [comment]
@ahardin-rh LGTM
Thank you for all efforts!

**Comment by ahardin-rh** [comment]
Thank you for the great reviews!

**Pull request merged**